probe-rs-cli-util/src/flash.rs

```diff
@@ -0,0 +1,424 @@
+        path: String,
+    },
+    #[error("Failed to change the working directory to '{path}'.")]
+    FailedToChangeWorkingDirectory {
+        #[source]
+        source: std::io::Error,
+        path: String,
+    },
+    #[error("Failed to build the cargo project at '{path}'.")]
+    FailedToBuildExternalCargoProject {
+        #[source]
+        source: ArtifactError,
+        path: String,
+    },
+    #[error("Failed to build the cargo project.")]
+    FailedToBuildCargoProject(#[source] ArtifactError),
+    #[error("The chip '{name}' was not found in the database.")]
+    ChipNotFound {
+        #[source]
+        source: RegistryError,
+        name: String,
+    },
+    #[error("The protocol '{protocol}' could not be selected.")]
+    FailedToSelectProtocol {
+        #[source]
+        source: DebugProbeError,
+        protocol: WireProtocol,
+    },
+    #[error("The protocol speed coudl not be set to '{speed}' kHz.")]
+    FailedToSelectProtocolSpeed {
+        #[source]
+        source: DebugProbeError,
+        speed: u32,
+    },
+    #[error("Connecting to the chip was unsuccessful.")]
+    AttachingFailed {
+        #[source]
+        source: probe_rs::Error,
+        connect_under_reset: bool,
+    },
+    #[error("Failed to get a handle to the first core.")]
+    AttachingToCoreFailed(#[source] probe_rs::Error),
+    #[error("The reset of the target failed.")]
+    TargetResetFailed(#[source] probe_rs::Error),
+    #[error("The target could not be reset and halted.")]
+    TargetResetHaltFailed(#[source] probe_rs::Error),
+    #[error("Failed to write to file")]
+    IOError(#[source] std::io::Error),
+}
+
+impl From<std::io::Error> for CargoFlashError {
+    fn from(e: std::io::Error) -> Self {
+        CargoFlashError::IOError(e)
+    }
+}
+
+/// Lists all connected debug probes.
+pub fn list_connected_probes(mut f: impl Write) -> Result<(), std::io::Error> {
+    let probes = Probe::list_all();
+
+    if !probes.is_empty() {
+        writeln!(f, "The following debug probes were found:")?;
+        for (num, link) in probes.iter().enumerate() {
+            writeln!(f, "[{}]: {:?}", num, link)?;
+        }
+    } else {
+        writeln!(f, "No debug probes were found.")?;
+    }
+
+    Ok(())
+}
+
+/// Print all the available families and their contained chips to the
+/// commandline.
+pub fn print_families(mut f: impl Write) -> Result<(), CargoFlashError> {
+    writeln!(f, "Available chips:")?;
+    for family in probe_rs::config::families().map_err(CargoFlashError::FailedToReadFamilies)? {
+        writeln!(f, "{}", &family.name)?;
+        writeln!(f, "    Variants:")?;
+        for variant in family.variants() {
+            writeln!(f, "        {}", variant.name)?;
+        }
+    }
+    Ok(())
+}
+
+impl FlashOptions {
+    /// Add targets contained in file given by --chip-description-path
+    /// to probe-rs registery.
+    pub fn try_load_chip_desc(&self) -> Result<(), CargoFlashError> {
```

Comment:
Perhaps `maybe_add_chip_descs` is more suitable, or otherwise some general `init`?

Reply:
Now named `maybe_*`.